auto_compressor/src/state_saving.rs

```diff
@@ -0,0 +1,260 @@
+        prev_seen = level_num;
+    }
+
+    // If we didn't retrieve anything from the database then there is no saved state
+    // in the database!
+    if level_info.is_empty() {
+        return Ok(None);
+    }
+
+    // Return the compressor state we retrieved
+    Ok(Some((last_compressed, level_info)))
+}
+
+/// Save the level info so it can be loaded by the next run of the compressor
+///
+/// # Arguments
+///
+/// * `client`            - A postgres client used to send the requests to the database
+/// * `room_id`           - The room who's saved compressor state we want to save
+/// * `level_info`        - The state that can be used to restore the compressor later
+/// * `last_compressed`   - The last state_group that was compressed. This is needed
+///                         so that the compressor knows where to start from next
+pub fn write_room_compressor_state(
+    client: &mut Client,
+    room_id: &str,
+    level_info: &[Level],
+    last_compressed: i64,
+) -> Result<()> {
+    // The query we are going to build up
+    let mut sql = "".to_string();
+
+    // Go through every level that the compressor is using
+    for (level_num, level) in level_info.iter().enumerate() {
+        // the 1st level is level 1 not level 0, but enumerate starts at 0
+        // so need to add 1 to get correct number
+        let level_num = level_num + 1;
+
+        // delete the old information for this level from the database
+        sql.push_str(&format!(
+            "DELETE FROM state_compressor_state WHERE room_id = {} AND level_num = {};\n",
+            PGEscape(room_id),
+            level_num,
+        ));
+
+        // bring the level info out of the tuple
+        let (max_size, current_len, current_head) = (
+            level.get_max_length(),
+            level.get_current_length(),
+            level.get_head(),
+        );
+
+        // Current_head is either a value or NULL
+        // need to convert from Option so that this can be placed into a string
+        let current_head = match current_head {
+            Some(s) => s.to_string(),
+            None => "NULL".to_string(),
+        };
+
+        // Add the new informaton for this level into the database
+        sql.push_str(&format!(
+            concat!(
+                "INSERT INTO state_compressor_state (room_id, level_num, max_size,",
+                " current_length, current_head) VALUES ({}, {}, {}, {}, {});"
+            ),
+            PGEscape(room_id),
+            level_num,
+            max_size,
+            current_len,
+            current_head,
+        ));
+    }
+
+    // delete the old information for this level from the database
+    sql.push_str(&format!(
+        "DELETE FROM state_compressor_progress WHERE room_id = {};\n",
+        PGEscape(room_id),
+    ));
+
+    // Add the new informaton for this level into the database
+    sql.push_str(&format!(
+        "INSERT INTO state_compressor_progress (room_id, last_compressed) VALUES ({},{});",
```

Comment:
You can replace the delete and insert with an upsert:

```sql
INSERT INTO state_compressor_progress (room_id, last_compressed)
VALUES ($1, $2)
ON CONFLICT (room_id) DO UPDATE SET last_compressed = EXCLUDED.last_compressed
```

and similarly above.

c.f. https://www.postgresql.org/docs/current/sql-insert.html